### PR TITLE
Refactored category information to come from backend

### PR DESCRIPTION
**Pull request by joeyballentine**
This accomplishes a few things:
- The category names used by nodes are now 100% guaranteed to be correct and matching to the actual categories 
- The category icons and colors are no longer hardcoded in the frontend as they are now defined in the backend. The only caveat to this is that custom icons (PyTorch, NCNN, and ONNX) still have to be defined in the frontend. 
- The install hints are also no longer hardcoded in the frontend

### CODE REVIEW

File: backend/src/nodes/categories.py
Comment (RunDevelopment) [inline backend/src/nodes/categories.py:24]: Please type the install hint as nullable instead of using the empty string as sentinel value.

File: src/common/common-types.ts
Comment (RunDevelopment) [inline src/common/common-types.ts:159]: This type also need to be nullable.

File: src/renderer/components/NodeSelectorPanel/NodeSelectorPanel.tsx
Comment (RunDevelopment) [inline src/renderer/components/NodeSelectorPanel/NodeSelectorPanel.tsx:225]: TODO: Open issue instead of using a TODO.

---

Yeah, I think this should be in issue rather than a TODO comment.

**Review by RunDevelopment — CHANGES_REQUESTED** [review]
I think the general approach is good. Just a few nits and suggestions.

> The category names used by nodes are now 100% guaranteed to be correct and matching to the actual categories

No, they aren't because all nodes use these constants and not `MyCategory.name`. Those constants just so happen to be used by categories.

IMO, there shouldn't be any category name constants. This will require more code changes, though.

> The category [...] colors [...] are now defined in the backend.

I don't like this. Color is such a pure frontend thing. What if light and dark mode require slightly different category colors in the future? What if we want to add themes that can also change category colors (e.g. a color-blind or high-contrast mode)? It's just not practical IMO.

**Comment by joeyballentine** [comment]
> Color is such a pure frontend thing. What if light and dark mode require slightly different category colors in the future?

I started typing up a big rebuttal to this, which I'll still post if you disagree with this idea, but I had an idea that's kind of a middle ground:

We define a color system similar to chakra's, where you use a string of a pre-defined color variable for the category (in the backend) like "category.red.500". Then, have frontend code to take that value and apply an actual color. I would be completely ok with doing this (in a separate PR ideally). What I'm not ok with is keeping the big hardcoded switch statement for category information.

**Comment by RunDevelopment** [comment]
I like the idea of category color variables in the frontend, but I think e.g. "category.red.500" is the wrong approach. This will just lead to a bunch of colors that aren't used (i.e. I doubt we'll use all "category.red.X00" colors). IMO we should just have something like CSS variables for our categories.

Anyway, this is nothing that should block this PR, so let's continue this discussion in an issue after merging this PR.

File: backend/src/nodes/image_adj_nodes.py
Comment (RunDevelopment) [inline backend/src/nodes/image_adj_nodes.py:41]: Looking at this, I have another idea: Let's make it `self.category = ImageAdjustmentCategory`.

This would make `self.category` an actual category object and not just some string, and we could save the `.name` everywhere. We would just have to change the type of `self.category` and make it `"category": node_object.get_category().name` in `run.py`.

What do you think?

Comment (joeyballentine) [reply backend/src/nodes/image_adj_nodes.py:41]: Sure

**Comment by joeyballentine** [comment]
> This will just lead to a bunch of colors that aren't used

I do not want to hardcode any of the categories in the frontend anywhere, period. I know you see it differently, but I want chainner to move in the direction of being able to use any backend with the same frontend. As if the frontend was a closed source project and we were developing a backend for it. That would mean we can't rely on changing the frontend to pick node colors, and would instead mean that the frontend provides essentially an API for how the backend picks what color a node is. Does that make sense?

**Review by RunDevelopment — APPROVED** [review]

**Pull request merged**
